File: NiftyOptionCalculator-Android/app/src/main/java/com/niftyoption/calculator/data/models/OptionData.kt
package com.niftyoption.calculator.data.models

import com.google.gson.annotations.SerializedName
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.UUID
import kotlin.math.abs
import kotlin.math.max

// MARK: - Option Type Enum

enum class OptionType(val code: String, val displayName: String) {
    CALL("CE", "Call"),
    PUT("PE", "Put");

    companion object {
        fun fromCode(code: String): OptionType = when (code) {
            "CE" -> CALL
            "PE" -> PUT
            else -> CALL
        }
    }
}

// MARK: - Option Data Model

data class OptionData(
    val id: String = UUID.randomUUID().toString(),
    val strikePrice: Double,
    val optionType: OptionType,
    val expiryDate: Date,
    val lastTradedPrice: Double,
    val openInterest: Int = 0,
    val changeInOI: Int = 0,
    val impliedVolatility: Double,
    val bidPrice: Double = 0.0,
    val askPrice: Double = 0.0,
    val bidQty: Int = 0,
    val askQty: Int = 0,
    val volume: Int = 0,
    val underlyingValue: Double
) {
    val isITM: Boolean
        get() = when (optionType) {
            OptionType.CALL -> underlyingValue > strikePrice
            OptionType.PUT -> underlyingValue < strikePrice
        }

    val isATM: Boolean
        get() = abs(underlyingValue - strikePrice) <= 50

    val moneyness: String
        get() = when {
            isATM -> "ATM"
            isITM -> "ITM"
            else -> "OTM"
        }

    val daysToExpiry: Int
        get() {
            val diff = expiryDate.time - Date().time
            return max((diff / (24 * 60 * 60 * 1000)).toInt(), 0)
        }

    val timeToExpiryYears: Double
        get() = daysToExpiry.toDouble() / 365.0

    val displayStrike: String
        get() = "%.0f".format(strikePrice)

    val displayLTP: String
        get() = "%.2f".format(lastTradedPrice)

    val displayIV: String
        get() = "%.2f%%".format(impliedVolatility * 100)
}

// MARK: - Option Chain Row

data class OptionChainRow(
    val id: String = UUID.randomUUID().toString(),
    val strikePrice: Double,
    val callOption: OptionData?,
    val putOption: OptionData?
) {
    val displayStrike: String
        get() = "%.0f".format(strikePrice)
}

// MARK: - Greeks Result

data class GreeksResult(
    val delta: Double,
    val gamma: Double,
    val theta: Double,  // Per day
    val vega: Double,   // Per 1% IV change
    val rho: Double     // Per 1% rate change
) {
    val displayDelta: String get() = "%.4f".format(delta)
    val displayGamma: String get() = "%.6f".format(gamma)
    val displayTheta: String get() = "%.2f".format(theta)
    val displayVega: String get() = "%.2f".format(vega)
    val displayRho: String get() = "%.2f".format(rho)

    val deltaInterpretation: String
        get() {
            val absDelta = abs(delta)
            return when {
                absDelta > 0.7 -> "Deep ITM - moves almost 1:1 with underlying"
                absDelta in 0.45..0.55 -> "ATM - 50% chance of expiring ITM"
                absDelta < 0.3 -> "Deep OTM - low probability of profit"
                else -> "Moderate sensitivity to underlying movement"
            }
        }

    val thetaInterpretation: String
        get() = when {
            theta < -5 -> "High time decay - losing value rapidly"
            theta < -2 -> "Moderate time decay"
            else -> "Low time decay"
        }

    val vegaInterpretation: String
        get() = when {
            vega > 10 -> "High volatility sensitivity"
            vega > 5 -> "Moderate volatility sensitivity"
            else -> "Low volatility sensitivity"
        }

    fun expectedPriceChange(underlyingMove: Double): Double {
        val firstOrder = delta * underlyingMove
        val secondOrder = 0.5 * gamma * underlyingMove * underlyingMove
        return firstOrder + secondOrder
    }

    fun expectedPrice(currentPrice: Double, underlyingMove: Double, daysElapsed: Double = 0.0): Double {
        val priceChange = expectedPriceChange(underlyingMove)
        val timeDecay = theta * daysElapsed
        return max(currentPrice + priceChange + timeDecay, 0.0)
    }
}

// MARK: - Target Calculation Result

data class TargetCalculation(
    val id: String = UUID.randomUUID().toString(),
    val optionType: OptionType,
    val strikePrice: Double,
    val currentSpot: Double,
    val currentOptionPrice: Double,
    val targetSpot: Double,
    val stopLossSpot: Double,
    val targetOptionPrice: Double,
    val stopLossOptionPrice: Double,
    val greeks: GreeksResult,
    val impliedVolatility: Double,
    val daysToExpiry: Int
) {
    val targetProfit: Double get() = targetOptionPrice - currentOptionPrice
    val stopLossLoss: Double get() = currentOptionPrice - stopLossOptionPrice
    val riskRewardRatio: Double get() = if (stopLossLoss > 0) targetProfit / stopLossLoss else 0.0

    val displayTargetPrice: String get() = "%.2f".format(targetOptionPrice)
    val displayStopLossPrice: String get() = "%.2f".format(stopLossOptionPrice)
    val displayProfit: String get() = "%+.2f".format(targetProfit)
    val displayLoss: String get() = "-%.2f".format(stopLossLoss)
    val displayRiskReward: String get() = "1:%.2f".format(riskRewardRatio)

    val profitPercentage: Double
        get() = if (currentOptionPrice > 0) (targetProfit / currentOptionPrice) * 100 else 0.0

    val lossPercentage: Double
        get() = if (currentOptionPrice > 0) (stopLossLoss / currentOptionPrice) * 100 else 0.0

    val displayProfitPercentage: String get() = "+%.1f%%".format(profitPercentage)
    val displayLossPercentage: String get() = "-%.1f%%".format(lossPercentage)
    val isGoodRiskReward: Boolean get() = riskRewardRatio >= 1.5
}

// MARK: - NSE API Response Models

data class NSEOptionChainResponse(
    val records: Records
) {
    data class Records(
        val expiryDates: List<String>,
        val data: List<DataItem>,
        val underlyingValue: Double,
        val strikePrices: List<Double>,
        val timestamp: String
    )

    data class DataItem(
        val strikePrice: Double,
        val expiryDate: String,
        @SerializedName("CE") val ce: OptionDetail?,
        @SerializedName("PE") val pe: OptionDetail?
    )

    data class OptionDetail(
        val strikePrice: Double,
        val expiryDate: String,
        val underlying: String,
        val identifier: String,
        val openInterest: Int,
        val changeinOpenInterest: Int,
        val pchangeinOpenInterest: Double,
        val totalTradedVolume: Int,
        val impliedVolatility: Double,
        val lastPrice: Double,
        val change: Double,
        val pChange: Double,
        val totalBuyQuantity: Int,
        val totalSellQuantity: Int,
        val bidQty: Int,
        val bidprice: Double,
        val askQty: Int,
        val askPrice: Double,
        val underlyingValue: Double
    )
}

// MARK: - Expiry Date Helper

data class ExpiryDate(
    val id: String,
    val date: Date,
    val displayString: String
) {
    companion object {
        fun fromString(dateString: String): ExpiryDate {
            val formatter = SimpleDateFormat("dd-MMM-yyyy", Locale("en", "IN"))
            val date = try {
                formatter.parse(dateString) ?: Date()
            } catch (e: Exception) {
                Date()
            }

            val displayFormatter = SimpleDateFormat("dd MMM", Locale.getDefault())
            val displayString = displayFormatter.format(date)

            return ExpiryDate(id = dateString, date = date, displayString = displayString)
        }
    }
}

// MARK: - UI State

sealed class LoadingState {
    data object Idle : LoadingState()
    data object Loading : LoadingState()
    data object Loaded : LoadingState()
    data class Error(val message: String) : LoadingState()
}

File: NiftyOptionCalculator-iOS/NiftyOptionCalculator/Models/OptionData.swift
import Foundation

// MARK: - Trading Index Model

enum TradingIndex: String, CaseIterable, Identifiable {
    case nifty50 = "NIFTY"
    case bankNifty = "BANKNIFTY"
    case niftyFinService = "FINNIFTY"
    case midcapSelect = "MIDCPNIFTY"
    case sensex = "SENSEX"
    case bankex = "BANKEX"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .nifty50: return "NIFTY 50"
        case .bankNifty: return "BANK NIFTY"
        case .niftyFinService: return "FIN NIFTY"
        case .midcapSelect: return "MIDCAP SELECT"
        case .sensex: return "SENSEX"
        case .bankex: return "BANKEX"
        }
    }

    var shortName: String {
        switch self {
        case .nifty50: return "NIFTY"
        case .bankNifty: return "BANKNIFTY"
        case .niftyFinService: return "FINNIFTY"
        case .midcapSelect: return "MIDCAP"
        case .sensex: return "SENSEX"
        case .bankex: return "BANKEX"
        }
    }

    var exchange: String {
        switch self {
        case .sensex, .bankex: return "BSE"
        default: return "NSE"
        }
    }

    // Upstox instrument key format
    var instrumentKey: String {
        switch self {
        case .nifty50: return "NSE_INDEX|Nifty 50"
        case .bankNifty: return "NSE_INDEX|Nifty Bank"
        case .niftyFinService: return "NSE_INDEX|Nifty Fin Service"
        case .midcapSelect: return "NSE_INDEX|NIFTY MID SELECT"
        case .sensex: return "BSE_INDEX|SENSEX"
        case .bankex: return "BSE_INDEX|BANKEX"
        }
    }

    // Strike price interval
    var strikeInterval: Double {
        switch self {
        case .nifty50: return 50
        case .bankNifty: return 100
        case .niftyFinService: return 50
        case .midcapSelect: return 25
        case .sensex: return 100
        case .bankex: return 100
        }
    }

    // Lot size (revised Nov 2024)
    var lotSize: Int {
        switch self {
        case .nifty50: return 75
        case .bankNifty: return 30
        case .niftyFinService: return 25
        case .midcapSelect: return 50
        case .sensex: return 10
        case .bankex: return 15
        }
    }

    // Expiry day of week (0 = Sunday, 4 = Thursday, 5 = Friday)
    var expiryDayOfWeek: Int {
        switch self {
        case .nifty50: return 4 // Thursday
        case .bankNifty: return 3 // Wednesday
        case .niftyFinService: return 2 // Tuesday
        case .midcapSelect: return 1 // Monday
        case .sensex: return 5 // Friday
        case .bankex: return 1 // Monday
        }
    }

    // Color for UI
    var themeColor: String {
        switch self {
        case .nifty50: return "00C805"  // Green
        case .bankNifty: return "007AFF" // Blue
        case .niftyFinService: return "FF9500" // Orange
        case .midcapSelect: return "AF52DE" // Purple
        case .sensex: return "FF3B30" // Red
        case .bankex: return "5AC8FA" // Cyan
        }
    }

    // Icon for UI
    var icon: String {
        switch self {
        case .nifty50: return "chart.line.uptrend.xyaxis"
        case .bankNifty: return "building.columns.fill"
        case .niftyFinService: return "indianrupeesign.circle.fill"
        case .midcapSelect: return "chart.bar.fill"
        case .sensex: return "s.circle.fill"
        case .bankex: return "b.circle.fill"
        }
    }

    // Approximate base price for mock data generation (Jan 2025)
    var basePrice: Double {
        switch self {
        case .nifty50: return 23500
        case .bankNifty: return 49500
        case .niftyFinService: return 23200
        case .midcapSelect: return 12500
        case .sensex: return 77500
        case .bankex: return 54000
        }
    }
}

// MARK: - Option Type Enum

enum OptionType: String, Codable, CaseIterable {
    case call = "CE"
    case put = "PE"

    var displayName: String {
        switch self {
        case .call: return L.calculatorCall
        case .put: return L.calculatorPut
        }
    }

    var shortName: String {
        return rawValue
    }
}

// MARK: - Option Data Model

struct OptionData: Identifiable, Codable, Equatable {
    let id: UUID
    let strikePrice: Double
    let optionType: OptionType
    let expiryDate: Date
    let lastTradedPrice: Double
    let openInterest: Int
    let changeInOI: Int
    let impliedVolatility: Double
    let bidPrice: Double
    let askPrice: Double
    let bidQty: Int
    let askQty: Int
    let volume: Int
    let underlyingValue: Double

    // Greeks from API (optional - if not provided, calculate using Black-Scholes)
    let delta: Double?
    let gamma: Double?
    let theta: Double?
    let vega: Double?

    // Upstox instrument key for WebSocket subscription
    let instrumentKey: String?

    init(
        id: UUID = UUID(),
        strikePrice: Double,
        optionType: OptionType,
        expiryDate: Date,
        lastTradedPrice: Double,
        openInterest: Int = 0,
        changeInOI: Int = 0,
        impliedVolatility: Double,
        bidPrice: Double = 0,
        askPrice: Double = 0,
        bidQty: Int = 0,
        askQty: Int = 0,
        volume: Int = 0,
        underlyingValue: Double,
        delta: Double? = nil,
        gamma: Double? = nil,
        theta: Double? = nil,
        vega: Double? = nil,
        instrumentKey: String? = nil
    ) {
        self.id = id
        self.strikePrice = strikePrice
        self.optionType = optionType
        self.expiryDate = expiryDate
        self.lastTradedPrice = lastTradedPrice
        self.openInterest = openInterest
        self.changeInOI = changeInOI
        self.impliedVolatility = impliedVolatility
        self.bidPrice = bidPrice
        self.askPrice = askPrice
        self.bidQty = bidQty
        self.askQty = askQty
        self.volume = volume
        self.underlyingValue = underlyingValue
        self.delta = delta
        self.gamma = gamma
        self.theta = theta
        self.vega = vega
        self.instrumentKey = instrumentKey
    }

    var isITM: Bool {
        switch optionType {
        case .call:
            return underlyingValue > strikePrice
        case .put:
            return underlyingValue < strikePrice
        }
    }

    var isATM: Bool {
        let diff = abs(underlyingValue - strikePrice)
        return diff <= 50 // Within 50 points considered ATM
    }

    var moneyness: String {
        if isATM { return "ATM" }
        return isITM ? "ITM" : "OTM"
    }

    var daysToExpiry: Int {
        let calendar = Calendar.current
        let components = calendar.dateComponents([.day], from: Date(), to: expiryDate)
        return max(components.day ?? 0, 0)
    }

    var timeToExpiryYears: Double {
        return Double(daysToExpiry) / 365.0
    }

    /// Precise time to expiry in years using seconds-based calculation.
    /// Accounts for NSE market close at 15:30 IST on expiry day,
    /// avoiding integer-day truncation that significantly misprices
    /// short-dated options.
    var preciseTimeToExpiryYears: Double {
        var calendar = Calendar.current
        calendar.timeZone = TimeZone(identifier: "Asia/Kolkata")!
        var components = calendar.dateComponents(in: calendar.timeZone, from: expiryDate)
        components.hour = 15
        components.minute = 30
        components.second = 0
        guard let adjustedExpiry = calendar.date(from: components) else {
            return timeToExpiryYears
        }
        let seconds = adjustedExpiry.timeIntervalSince(Date())
        return max(seconds / (365.25 * 24 * 3600), 0)
    }

    var displayStrike: String {
        return String(format: "%.0f", strikePrice)
    }

    var displayLTP: String {
        // Ensure we don't display negative prices (can happen with mock data)
        let safeLTP = max(lastTradedPrice, 0.05)
        return String(format: "%.2f", safeLTP)
    }

    var displayIV: String {
        return String(format: "%.2f%%", impliedVolatility * 100)
    }
}

// MARK: - Option Chain Strike Row

struct OptionChainRow: Identifiable {
    var id: Double { strikePrice }
    let strikePrice: Double
    let callOption: OptionData?
    let putOption: OptionData?

    init(strikePrice: Double, callOption: OptionData?, putOption: OptionData?) {
        self.strikePrice = strikePrice
        self.callOption = callOption
        self.putOption = putOption
    }

    var displayStrike: String {
        return String(format: "%.0f", strikePrice)
    }
}

// MARK: - Option Chain Response Model (NSE API)

struct NSEOptionChainResponse: Codable {
    let records: Records

    struct Records: Codable {
        let expiryDates: [String]
        let data: [DataItem]
        let underlyingValue: Double
        let strikePrices: [Double]
        let timestamp: String

        enum CodingKeys: String, CodingKey {
            case expiryDates
            case data
            case underlyingValue
            case strikePrices
            case timestamp
        }
    }

    struct DataItem: Codable {
        let strikePrice: Double
        let expiryDate: String
        let ce: OptionDetail?
        let pe: OptionDetail?

        enum CodingKeys: String, CodingKey {
            case strikePrice
            case expiryDate
            case ce = "CE"
            case pe = "PE"
        }
    }

    struct OptionDetail: Codable {
        let strikePrice: Double
        let expiryDate: String
        let underlying: String
        let identifier: String
        let openInterest: Int
        let changeinOpenInterest: Int
        let pchangeinOpenInterest: Double
        let totalTradedVolume: Int
        let impliedVolatility: Double
        let lastPrice: Double
        let change: Double
        let pChange: Double
        let totalBuyQuantity: Int
        let totalSellQuantity: Int
        let bidQty: Int
        let bidprice: Double
        let askQty: Int
        let askPrice: Double
        let underlyingValue: Double

        enum CodingKeys: String, CodingKey {
            case strikePrice
            case expiryDate
            case underlying
            case identifier
            case openInterest
            case changeinOpenInterest
            case pchangeinOpenInterest
            case totalTradedVolume
            case impliedVolatility
            case lastPrice
            case change
            case pChange
            case totalBuyQuantity
            case totalSellQuantity
            case bidQty
            case bidprice
            case askQty
            case askPrice
            case underlyingValue
        }
    }
}

// MARK: - Expiry Date Helper

struct ExpiryDate: Identifiable, Hashable {
    let id: String
    let date: Date
    let displayString: String

    init(dateString: String) {
        self.id = dateString

        let formatter = DateFormatter()
        formatter.dateFormat = "dd-MMM-yyyy"
        formatter.locale = Locale(identifier: "en_IN")

        if let date = formatter.date(from: dateString) {
            self.date = date
            let displayFormatter = DateFormatter()
            displayFormatter.dateFormat = "dd MMM"
            self.displayString = displayFormatter.string(from: date)
        } else {
            self.date = Date()
            self.displayString = dateString
        }
    }
}
